[HUDI-2417] Add support allowDuplicateInserts in HoodieJavaClient

dongkelun opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

*Add support allowDuplicateInserts in HoodieJavaClient*

## Brief change log

*(for example:)*
  - *Add support allowDuplicateInserts in HoodieJavaClient*

## Verify this pull request

This change added tests and can be verified as follows:

  - *Added testHoodieConcatHandle in TestJavaCopyOnWriteActionExecutor*


## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

pratyakshsharma commented:
LGTM!

vinothchandar commented:
cc @yihua 

#3696 is refactoring a bunch of this to ensure such cross client inconsistencies are minimized. Will take a look once we land that.

vinothchandar commented on hudi-client/hudi-java-client/src/test/java/org/apache/hudi/table/action/commit/TestJavaCopyOnWriteActionExecutor.java:
```diff
@@ -480,4 +478,76 @@ public void testBulkInsertRecords(String bulkInsertMode) throws Exception {
+    records2.add(new HoodieRecord(new HoodieKey(insertRow1.getRowKey(), insertRow1.getPartitionPath()), insertRow1));
+    records2.add(new HoodieRecord(new HoodieKey(insertRow2.getRowKey(), insertRow2.getPartitionPath()), insertRow2));
+
+    Thread.sleep(1000);
```
please avoid relying on sleep in tests. You can generate the commit times in the test, for greater control? Also could we move this test as a separate one for `HoodieConcatHandle`?

vinothchandar commented on hudi-client/hudi-java-client/src/test/java/org/apache/hudi/table/action/commit/TestJavaCopyOnWriteActionExecutor.java:
```diff
@@ -480,4 +478,76 @@ public void testBulkInsertRecords(String bulkInsertMode) throws Exception {
   }
+
+  @Test
+  public void testHoodieConcatHandle() throws Exception {
```
Wondeing if this can be rewritten more nicely using the TestTable abstraction. cc @xushiyan .

vinothchandar reviewed (commented):
Left some comments on the tests, but given the other PR is delayed. Might be okay to land this and reconcile. but still this is just duplicating code from the spark action executor

dongkelun commented:
@vinothchandar Hello, I have added comments, removed sleep and rebase with latest master, and put the test cases in separate class: `TestHoodieConcatHandle`.Can you take a look?

nsivabalan commented on hudi-client/hudi-java-client/src/test/java/org/apache/hudi/table/action/commit/TestHoodieConcatHandle.java:
```diff
@@ -0,0 +1,164 @@
+    int index = 0;
+    for (GenericRecord record : fileRecords) {
+      assertEquals(records1.get(index).getRecordKey(), record.get("_row_key").toString());
+      if (index == 3) {
```
can we also add assertion for index = 2. I assume "number" value should be 3.

dongkelun replied:
OK, I'll add it later

nsivabalan reviewed (changes requested):
changes in general is good. just 1 nit. Can we try to see if we can leverage HoodieTestDataGenerator to write tests. 
you can do something like this
dataGen.generateInserts : batch1
dataGen.generateUpdates : batch2
if not for allowDuplicates config set, total record count in snapshot view will be just batch1 records (assuming we updated all records from batch1)
if config is set to true, total record count in snapshot view will be batch1 records + batch2 records. and you can do some assertions to ensure both batch of records are present.

dongkelun commented:
> changes in general is good. just 1 nit. Can we try to see if we can leverage HoodieTestDataGenerator to write tests. you can do something like this dataGen.generateInserts : batch1 dataGen.generateUpdates : batch2 if not for allowDuplicates config set, total record count in snapshot view will be just batch1 records (assuming we updated all records from batch1) if config is set to true, total record count in snapshot view will be batch1 records + batch2 records. and you can do some assertions to ensure both batch of records are present.

OK, I'll try

nsivabalan commented:
since you have already put in time to write some tests, let already existing tests be there. may be in addition, give it a try.

nsivabalan commented on hudi-client/hudi-java-client/src/test/java/org/apache/hudi/table/action/commit/TestHoodieConcatHandle.java:
```diff
@@ -0,0 +1,226 @@
+
+    int index = 0;
+    for (GenericRecord record : fileRecords) {
+      assertEquals(records1.get(index).getRecordKey(), record.get("_row_key").toString());
```
I see records1 stores the expected output. incase of mergeAllowDups is false, should we expect records from records2 only?  as per this test code, we expect records from reords1. can you help me understand.

dongkelun replied:
If mergeAllowDups is true, records1 will traverse completely. If it is false, it will only traverse half. We can also traverse records2 instead. For the sake of code simplicity, I didn't change it to record2.Do I need to change to record2?

dongkelun replied:
The "_row_key"  of records1 and record2 are the same, so when mergeallowdups is false, the effect of judging "_row_key" is the same
